torch/onnx/_internal/fx/function_dispatcher.py

```diff
@@ -3,306 +3,333 @@
-                continue
-
-            exporter_look_up_key = op_overload_packet._qualified_op_name
-            if _ATENLIB_FUNCTIONS.get(exporter_look_up_key) is None:
-                # This aten op doesn't have ONNX exporter.
-                continue
-
-            for overload_name in op_overload_packet.overloads():
-                op_overload = getattr(op_overload_packet, overload_name)
-                # This line maps torch.ops.aten.add.Tensor, torch.ops.aten.add.Scalar, torch.ops.aten.add.out, etc
-                # to "aten::add". This means the exporter for "aten::add" is used for all overloads of "aten::add".
-                # This is applied to all ops under torch.ops.aten.
-                #
-                # TODO(wechi): in the future, we might want to write individual exporter for each overload, if,
-                # for example, they have different type promotion rules. If so, just map different overloads to
-                # different exporter keys.
-
-                table[op_overload] = op_overload_packet._qualified_op_name
-    return table
-
-
-# Dictionary that maps torch.ops.aten.* to exporter look up key; e.g.,
-# _OP_OVERLOAD_TO_EXPORTER_KEY_TABLE[torch.add.Tensor] is "aten::add".
-_OP_OVERLOAD_TO_EXPORTER_KEY_TABLE = _create_op_overload_to_exporter_key_table()
-_SYMINT_SYMFLOAT_BUILTIN_TO_EXPORTER_KEY_TABLE = {
-    operator.mul: "aten::mul",
-    operator.add: "aten::add",
-    operator.pow: "aten::pow",
-    operator.sub: "aten::sub",
-}
+class OnnxDispatcher:
+    """
+    The OnnxDispatcher class finds the best matched function for a given aten operation
+    in the FX exporter. It uses the torch.ops name to find the function. If not found,
+    it falls back to default. Then, it finds the best match among all overloaded
+    functions. If the types match, it selects the function. Otherwise, it finds the
+    nearest upcast/downcast: Float to Float / Int to Int, Int to Float.
 
+    Steps for overloaded function dispatch:
 
-@_beartype.beartype
-def _create_onnx_friendly_decomposition_table() -> (
-    Dict[torch._ops.OpOverload, Callable]
-):
-    decomposition_table: Dict[torch._ops.OpOverload, Callable] = {}
-    for op_overload, decomp_fn in torch._decomp.decomposition_table.items():
-        # Skip decomposition into "prim::*" ops (defined in 'torch._refs'), because they
-        # are not generally supported by ONNX.
-        # Skip decomposition for op_overload as long as that op_overload has a corresponding ONNX
-        # symbolic function.
-        if (
-            "torch._refs" in decomp_fn.__module__
-            or op_overload in _OP_OVERLOAD_TO_EXPORTER_KEY_TABLE
+    1. Use the torch.ops name to find the function:
+        a. Check if the ATen overload exists.
+        b. If not found, check ATen overload=default.
+
+    2. Find the best match among all overloaded functions:
+        a. If the types match, select the function.
+    """
```

Comment:
If there is not perfect match, we choose the most matched one.

Reply:
Having this captured by SARIF is fine, but it should not replace docstring documentation. docstring help us compare the actual behavior from the expected behavior. SARIF only helps with the former